packages/google_maps_flutter/google_maps_flutter_web/lib/src/convert.dart

```diff
@@ -262,20 +271,178 @@ gmaps.Size? _gmSizeFromIconConfig(List<Object?> iconConfig, int sizeIndex) {
+/// result of the size calculation to improve performance.
+///
+/// [icon] is the [gmaps.Icon] object whose size is being set.
+///
+/// This method asserts that [url] is not empty when [width] and [height] are
+/// null, ensuring that a URL is provided for fetching the image.
+Future<void> _setIconSize({
+  required double? width,
+  required double? height,
+  required double imagePixelRatio,
+  required String url,
+  required String cacheKey,
+  required gmaps.Icon icon,
+}) async {
+  if (width != null && height != null) {
+    final gmaps.Size gmapsSize = gmaps.Size(width, height);
+    icon.size = gmapsSize;
+    icon.scaledSize = gmapsSize;
+  } else {
+    assert(
+        url.isNotEmpty, 'URL must not be empty when calculating dimensions.');
+    final Size? bitmapSize = await _getBitmapSize(url, cacheKey);
+
+    // If bitmap size is null, the image is invalid,
+    // and the icon size is not set.
+    if (bitmapSize != null) {
+      double targetWidth = bitmapSize.width;
+      double targetHeight = bitmapSize.height;
+
+      if (width == null && height == null) {
+        // Adjust the target size based on the image pixel ratio.
+        targetWidth /= imagePixelRatio;
+        targetHeight /= imagePixelRatio;
+      } else {
+        final double aspectRatio = bitmapSize.width / bitmapSize.height;
+        targetWidth = width ?? (height ?? bitmapSize.height) * aspectRatio;
+        targetHeight = height ?? (width ?? bitmapSize.width) / aspectRatio;
+      }
 
+      // Set the icon sizes.
+      icon.size = gmaps.Size(targetWidth, targetHeight);
+      icon.scaledSize = gmaps.Size(targetWidth, targetHeight);
+    }
+  }
+}
+
+/// Fetches the size of the bitmap from a given URL and caches the result.
+///
+/// This method attempts to fetch the image size for a given [url] and caches
+/// the result using [cacheKey]. If the size has already been calculated and
+/// cached for this key, the cached value is returned immediately to improve
+/// performance.
+///
+/// - Parameters:
+/// [url] is the URL of the image for which the size is being fetched.
+/// [cacheKey] is a unique key for each image used for caching the size calculation result.
+Future<Size?> _getBitmapSize(String url, String cacheKey) async {
+  if (_bitmapSizeCompleterCache.containsKey(cacheKey)) {
+    return _bitmapSizeCompleterCache[cacheKey]!.future;
+  }
+  final Completer<Size?> completer = Completer<Size?>();
+  _bitmapSizeCompleterCache[cacheKey] = completer;
+  final HTMLImageElement image = HTMLImageElement()..src = url;
+
+  // Wait for the onLoad or onError event.
+  await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);
+
+  if (image.width == 0 || image.height == 0) {
+    // Complete with null for invalid images.
+    completer.complete(null);
+  } else {
+    // Complete with the image size for valid images.
+    completer.complete(Size(image.width.toDouble(), image.height.toDouble()));
+  }
```

Comment:
This method should be equivalent to yours, I think:

```dart
// _bitmapSizeCache should be a Map<String, Size?>

Future<Size?> _fetchBitmapSize(String url) async {
  if (!_bitmapSizeCache.containsKey(url)) {
    final HTMLImageElement image = HTMLImageElement()..src = url;

    // Wait for the onLoad or onError event.
    await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);

    if (image.width == 0 || image.height == 0) {
      // Complete with null for invalid images.
      _bitmapSizeCache[url] = null;
    } else {
      // Complete with the image size for valid images.
      _bitmapSizeCache[url] = Size(image.width.toDouble(), image.height.toDouble());
    }
  }
  return _bitmapSizeCache[url]; // May be null
}
```

By making the method `async` and using an `await` in the middle of it, you already have a "free completer" that will be fulfilled when you `return`.

Reply:
This is not a same. 
As _fetchBitmapSize can be called multiple times before 
`await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first])` is resolved and the resolved size is cached with url.
This is because multiple markers are handled in paraller for speed.

Reply:
This would work:

```dart
final Map<String, Future<Size?>> _bitmapSizeFutureCache =
    <String, Future<Size?>>{};

...

/// Fetches the size of the bitmap from a given URL and caches the result.
///
/// This method attempts to fetch the image size for a given [url].
Future<Size?> _fetchBitmapSize(String url) async {
  final HTMLImageElement image = HTMLImageElement()..src = url;

  // Wait for the onLoad or onError event.
  await Future.any(<Future<Event>>[image.onLoad.first, image.onError.first]);

  if (image.width == 0 || image.height == 0) {
    // Complete with null for invalid images.
    return null;
  }

  // Complete with the image size for valid images.
  return Size(image.width.toDouble(), image.height.toDouble());
}
```

And use this like this

```dart
    final Size? bitmapSize = await _bitmapSizeFutureCache.putIfAbsent(url, () {
      return _fetchBitmapSize(url);
    });
```

Reply:
Committed, as same future can be read multiple times, this works as well as completer, and is much simpler; thanks!